[k-hian] iP

k-hian opened the pull request:
# RoTodo
> RoTodo can't read your mind, otherwise RoTodo's creator would be rich! -RoTodo

RoTodo is a simple todo list manager. It's
- test-based 
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is:
1. Download from [here](https://github.com/k-hian/ip/releases/tag/A-Jar)
2. Run it
3. Add your tasks
4. Let RoTodo do the rest! :smile:

And it's **FREE**!

Features:
- [X] Manage tasks
- [ ] Manage deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class RoTodo {
    public static void main(String[] args) {
        RoTodo rotodo = new RoTodo();
        rotodo.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

k-hian committed:
Rename, add greeting and exit message

- renamed bot to "RoTodo"
- added greeting on running and exit message

k-hian committed:
Add input echo

- RoTodo echoes input back to user
- RoTodo exits program when receiving "bye" input

k-hian committed:
Implement "add task" and "list"

- user inputs are added as task
- "list" input will print out tasklist

k-hian committed:
Add "mark as done" and 3 task subtypes

- Task subclasses added: ToDos, Events, Deadlines

k-hian committed:
Automated Text UI Testing

k-hian committed:
Add exception handling

- Refactor code
   - move user input processing out of tasklist class into main CLI
     class (RoTodo)
   - organize codebase into packages:
      - rotodo.tasklist
      - rotodo.exception
- RoTodo can validate user input and outputs error message accordingly
   - custom exceptions added: InvalidInputException,
     IncompleteInputException

k-hian committed:
Add help command, update test-ui-testing

k-hian committed:
Add delete command

k-hian committed:
Add save and load data feature

RoTodo now loads data on startup and saves data if properly exited.

Refactor:
- RoTodo main class refactored into Parser, Storage and Ui to handle
  individual processes.
- saveString functions added to Tasks to format string for data saving.
- saveList function added to TaskList to facilitate data saving.

Todo:
- Modify text-ui-test to test load/save feature

k-hian committed:
Add comments for switch-case fallthrough

k-hian committed:
Merge branch 'branch-Level-7'

k-hian committed:
Add date/time parser

User input date/time is parsed into LocalDateTime format, but requires a
strict date/time format (dd/MM/yyyy HHmm).

Processes performed are not clearly segregated between Parser, TaskList
and Ui.

Commands that should be performed in the same scope is being performed
by different classes, affecting code readability.

Abstracting user commands into individual classes allow for better code
readability, and allow the processes to run in the same scope.

Let's define Command classes, that can be execute() to perform the
command as required.

k-hian committed:
Merge branch 'branch-Level-8'

k-hian committed:
Merge remote-tracking branch 'origin/add-gradle-support'

k-hian committed:
Add gradle support

Add javadoc to allow gradle to build with checkstyle.

k-hian committed:
Fix incorrect IncompleteInput check

k-hian committed:
Add JUnit tests

- TaskListTest for TaskList class
- AddCommandTest for AddCommand class

k-hian committed:
Add archiveVersion to build.gradle

k-hian committed:
Command.java: Add javadoc for execute method

k-hian committed:
Add find command

JUnit update:
- add find method test to TaskListTest

k-hian committed:
Follow coding standard

k-hian committed:
Follow the coding standard

k-hian committed:
Merge branch 'branch-A-CodingStandard'

k-hian committed:
Merge branch 'branch-Level-9'

k-hian committed:
Fix coding style

k-hian committed:
Merge branch 'branch-A-CheckStyle'

k-hian committed:
Add GUI using JavaFX

KengHian commented on src/main/java/rotodo/commands/AddCommand.java:
```diff
@@ -0,0 +1,122 @@
+    private String value;
+    private LocalDateTime from;
+    private LocalDateTime byto;
+    private boolean status;
```
I don't think this Boolean variable is descriptive enough and it should sound like Booleans as well (such as `isDone`) .

KengHian commented on src/main/java/rotodo/commands/AddCommand.java:
```diff
@@ -0,0 +1,122 @@
+    private TaskType type;
+    private String value;
+    private LocalDateTime from;
+    private LocalDateTime byto;
```
I think variable name should be in camelCase since it comprises of two words.

KengHian commented on src/main/java/rotodo/commands/AddCommand.java:
```diff
@@ -0,0 +1,122 @@
+                        "RoTodo can't read your mind, otherwise "
+                        + "RoTodo's creator would be rich!\n"
+                        + "  RoTodo needs a task description and deadline");
+                }
```
Great use of line break to improve readability! However, the indentation for wrapped lines should be 8 spaces.

KengHian commented on src/main/java/rotodo/commands/FindCommand.java:
```diff
@@ -0,0 +1,25 @@
+public class FindCommand extends Command {
+    private String keyword;
+
+    public FindCommand(String keyword) {
```
I think you should add a JavaDoc for this constructor as well

KengHian commented on src/main/java/rotodo/commands/MarkCommand.java:
```diff
@@ -0,0 +1,43 @@
+    public MarkCommand(int i, boolean as) {
+        idx = i;
+        asStatus = as;
+    }
```
Similar to a pervious comment, I think the naming of these Boolean variable could be change to sound like Booleans.

KengHian commented on src/main/java/rotodo/tasklist/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    /**
+     * Unmark the i'th Task as 'done'.
```
In method header comments, the first sentence should start in the form of `Unmarks...`.

KengHian commented on src/main/java/rotodo/tasklist/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @param value description of task
+     * @param by datetime
+     * @return
```
You did not specify what is being return here!

KengHian commented on src/main/java/rotodo/tasklist/Task.java:
```diff
@@ -0,0 +1,73 @@
+     */
+    public String unmarkAsDone() {
+        done = false;
+        return "Did something happen? RoTodo is confused... Task undone:\n  " + this.toString();
```
🥹

KengHian commented on src/main/java/rotodo/tasklist/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+    /**
+     * List of tasks.
+     */
+    private List<Task> list;
```
Perhaps a better variable name here would be `tasks`?

juliantayyc replied:
agreed ^

KengHian reviewed (commented):
Overall, I found your code is clear and well structure. However there are a few nips here and there as we are have to "just follow" the coding standard provided. I did not comment on all of the violations I came across, but broadly they are:

* `Boolean` variable naming
* Indentation for wrapped lines should be 8 spaces
* Some missing JavaDoc for public methods
* A few JavaDoc using the wrong tense

Outside of coding standard, I also think some of the variables' names are not immediately clear to me what they do, so perhaps is the something that could be improved upon. 

Lastly, good effort in giving RoTodo a personality!

juliantayyc commented on src/main/java/rotodo/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,30 @@
+ * @version CS2103T AY24/25 Semester 1
+ */
+public class DeleteCommand extends Command {
+    private int idx;
```
Perhaps you could consider using the full name spelt out instead of "texting style" spelling like we learnt this week. Just a nitpick though.

juliantayyc commented on src/main/java/rotodo/exception/InvalidInputException.java:
```diff
@@ -0,0 +1,41 @@
+public class InvalidInputException extends Exception {
+    private static int count = 0;
+    private static String[] error = new String[] {
+        "RoTodo.Error: Invalid Input",
```
HAHA nice!

juliantayyc commented on src/main/java/rotodo/processes/Ui.java:
```diff
@@ -0,0 +1,91 @@
+            + " R|____|_  /E \\____/ |___|  \\____/  \\_____/  \\_____/  \\____/  \\___/ (_) O o .\n"
+            + "        R\\/E\n").replace("R", ansiRed).replace("E", ansiReset)
+            + "Your very own Robot Todo List!\n"
+            + "How can I help you help yourself?";
```
woah... nice!

juliantayyc reviewed (commented):
LGTM, just some minor nitpicks. Your app is so unique! Well done!

k-hian committed:
Merge branch 'branch-Level-10'

k-hian committed:
Add javafx dependencies to build.gradle

k-hian committed:
Fix coding standards based on feedback

k-hian committed:
Add missing info to javadoc

`return` item in TaskList:addTask javadoc is left empty by mistake.

Add missing `return` details to javadoc for (Deadline) addTask.

k-hian committed:
Add assertions to commands

Added assertions to Command class execute methods. Ensures that required
objects are available for use by Command class.

Examples of added assertions:
- `assert ui != null;`
- `assert tl != null;`
- `assert st != null;`

Added assertions to RoTodo getResponse method to ensure input and
command are not null.

k-hian committed:
Improve code quality

k-hian committed:
Merge pull request #1 from k-hian/branch-A-Assertions

Add assertions to commands

k-hian committed:
Merge branch 'master' into branch-A-CodeQuality

k-hian committed:
Use streams

k-hian committed:
Merge pull request #2 from k-hian/branch-A-CodeQuality

Improve code quality

k-hian committed:
Merge branch 'master' into branch-A-Streams

k-hian committed:
Merge pull request #3 from k-hian/branch-A-Streams

Use streams

k-hian committed:
Set up CI

k-hian committed:
Fix find command

Bug 1: Find command does not find task that contain the keyword as part
of a longer word.

For e.g.
- Task description: "conCATenate"
- Keyword: "cat"
- Should identify task as found, but does not

Change `description.equals(keyword)` -> `description.contains(keyword)`.

Bug 2: If there is only one task that contains keyword (as part of a
longer word), then RoTodo will output no task found even if task was
found.

Add check for any search result.

k-hian committed:
Update runtest.bat

k-hian committed:
Update CI IO redirection text

k-hian committed:
Minor fixes to javaDoc and save file

Change save file directory from "data/" to ".data/".

Apply saveList on delete.

k-hian committed:
Add JUnit test for DeleteCommand

k-hian committed:
Fix saveList not running on modification

k-hian committed:
Add user guide

k-hian committed:
Format README

k-hian committed:
Fix JUnit test

k-hian committed:
Improve code quality

Change javadoc to better match phrasing standard of header.

Refactor `process` method in Parser class.

k-hian committed:
Fix README formating bug

k-hian committed:
Fix README table formatting

k-hian committed:
Fix wrong closing tags in README
